### PR TITLE
VM performance Tests Improvements 

### PR DESCRIPTION
**Pull request by SergioDemianLerner**
This is an improvement over the VM performance tests, with special emphasis in CALL performance. Asserts were added to many performance tests, even if these tests are ignored by default. The tests could be enabled if the unit test running environment is repeatable or they could be run in QA.

### CODE REVIEW

**Comment by tinchou** [comment]
1. I think this should be tackled by the QA team, setting goals and long term plans
2. I think we should use well-known tools such as `JMH - Java Microbenchmark Harness` instead of rolling our own measuring tools
3. I think that writing performance assertions that could make our CI builds fail because of transient conditions (e.g. the server is loaded) is a bad idea.

**Comment by SergioDemianLerner** [comment]
> 1. I think this should be tackled by the QA team, setting goals and long term plans
> 2. I think we should use well-known tools such as `JMH - Java Microbenchmark Harness` instead of rolling our own measuring tools
> 3. I think that writing performance assertions that could make our CI builds fail because of transient conditions (e.g. the server is loaded) is a bad idea.

1) I think RSK Labs has no QA team with enough resources to tackle any of these matters and the community needs tools to analyze VM performance right now. By preventing this merge your're preventing the RSK community to have visibility of the performance bottlenecks of the RSK platform. 

2) It doesn't matter whether you think this could be implemented using whatever better technology. There is always a better technology. As long as you don't implement it, basically your're just preventing progress. 

3) Those performance assertions are ignored by default (see @Ignore). Therefore they cannot make CI builds fail.

Last, I'm the only one testing VM performance  (you're not, clearly because you don't have the tools). So currently I'm the QA team over VM issues. I run these scripts frequently so I need them to be merged to detect unintended deterioration of VM performance (which happened in 2017, when the logging system was updated, and this went unnoticed until I ran the tests) .

**Comment by tinchou** [comment]
> I think RSK Labs has no QA team with enough resources to tackle any of these matters and the community needs tools to analyze VM performance right now. By preventing this merge your're preventing the RSK community to have visibility of the performance bottlenecks of the RSK platform.

I don't agree these `@Ignore`d tests provide visibility. They are not published, gathered periodically, or even documented.

> It doesn't matter whether you think this could be implemented using whatever better technology. There is always a better technology. As long as you don't implement it, basically your're just preventing progress.

As a core developer working on RSKj most of my day, I would argue that this kind of code is what prevents progress in the long run. When I'm trying to implement a feature, find a bug or refactor, I get slowed down by the myriad of undocumented tools, unused classes, methods for testing. Code is written once but read hundreds of times, and maintainability is key. You have faced these issues when working on the Unitrie!

> Those performance assertions are ignored by default (see @ignore). Therefore they cannot make CI builds fail.

Great, good to know.

> Last, I'm the only one testing VM performance (you're not, clearly because you don't have the tools). So currently I'm the QA team over VM issues. I run these scripts frequently so I need them to be merged to detect unintended deterioration of VM performance (which happened in 2017, when the logging system was updated, and this went unnoticed until I ran the tests).

You are the only one testing performance because we haven't prioritized that, thank you for covering us where we can't reach yet. But the Core team is the only one working on writing solid code. We are QA over maintainability issues.

I don't think that not having enough resources is a good reason not to do the best we can, and set up for growth and success. Even more, we wouldn't detect unintended deterioration of the VM if we don't run things manually. This proposal is not a solution, it's just convenience for one specific case but with a lot of drawbacks.

Let's plan for the long run!

**Comment by SergioDemianLerner** [comment]
> I don't think that not having enough resources is a good reason not to do the best we can,

That sentence is flawed in so many ways I can't count them. This is an adversarial scenario: people are going to find every possible way to attack RSK code, and flaws in VM opcodes is one of them. You don't wait to put a lock in your door because there is a better door that you will buy next year. You lock your door with whatever you have at hand to prevent attackers from entering. Then you go and query door vendors.

So the question is: are you going to start testing VM performance the next week ? if yes, then I will drop this PR.

File: rskj-core/src/test/java/co/rsk/vm/VMSpecificOpcodesPerformanceTest.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    @Ignore
+    @Test
+    public void testCallsToExistentAccounts() {
+        int gasCostPerCALLLoop = 755;
+        RskAddress shortAddress = new RskAddress("0000000000000000000000000000000000100000");
+        invoke.addAccount(shortAddress,new Coin(BigInteger.valueOf(1)));
+
+        // This code executes a CALL 256 times, to 256 different non-existing contracts
```
Comment (SergioDemianLerner) [inline rskj-core/src/test/java/co/rsk/vm/VMSpecificOpcodesPerformanceTest.java]: This comment is incorrect: it calls loopCount times (not 256)

File: rskj-core/src/test/java/co/rsk/vm/VMSpecificOpcodesPerformanceTest.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    private final TestSystemProperties config = new TestSystemProperties();
+    private final VmConfig vmConfig = config.getVmConfig();
+    private final PrecompiledContracts precompiledContracts = new PrecompiledContracts(config);
+
+    public static String padZeroesLeft(String s, int n) {
+        return StringUtils.leftPad(s, n, '0');
+    }
+
+    int loopCount = 0x1a00;
```
[the remark this answers is not in the document]
Comment (SergioDemianLerner) [reply rskj-core/src/test/java/co/rsk/vm/VMSpecificOpcodesPerformanceTest.java]: No, it's actually variable. The generated code takes it into account. What is incorrect is the comment that says "256 times". thanks

**Pull request merged**